src/main/java/frc/robot/subsystems/hood/HoodIOReal.java

```diff
@@ -0,0 +1,49 @@
+package frc.robot.subsystems.hood;
+
+import com.ctre.phoenix6.controls.DutyCycleOut;
+import com.ctre.phoenix6.controls.MotionMagicExpoTorqueCurrentFOC;
+import com.ctre.phoenix6.hardware.TalonFX;
+import edu.wpi.first.units.Angle;
+import edu.wpi.first.units.MutableMeasure;
+import edu.wpi.first.units.Units;
+import edu.wpi.first.wpilibj.DutyCycleEncoder;
+
+public class HoodIOReal implements HoodIO {
+    private final TalonFX motor;
+    private final DutyCycleEncoder absoluteEncoder = new DutyCycleEncoder(9); // TODO: to be changed
+    private final MotionMagicExpoTorqueCurrentFOC positionControl =
+            new MotionMagicExpoTorqueCurrentFOC(0);
+    private final DutyCycleOut dutyCycleOut = new DutyCycleOut(0);
+
+    public HoodIOReal() {
+        motor = new TalonFX(6); // TODO: to be changed
+        motor.getConfigurator().apply(HoodConstants.motorConfiguration);
+    }
+
+    @Override
+    public void updateInternalEncoder() {
+        motor.setPosition(absoluteEncoder.getAbsolutePosition());
+    }
+
+    @Override
+    public void setAngle(MutableMeasure<Angle> angle) {
+        inputs.controlMode = Mode.ANGLE;
+        inputs.angleSetpoint.mut_replace(angle);
+        motor.setControl(positionControl.withPosition(angle.in(Units.Rotations)));
+    }
+
+    @Override
+    public void setPower(double power) {
+        inputs.controlMode = Mode.POWER;
+        inputs.powerSetpoint = power;
+        motor.setControl(dutyCycleOut.withOutput(power));
+    }
+
+    @Override
+    public void updateInputs() {
+        inputs.angle.mut_replace(motor.getPosition().getValue(), Units.Rotations);
+        inputs.absoluteEncoderAngle.mut_replace(
+                absoluteEncoder.getAbsolutePosition(), Units.Rotations);
```

Comment:
The absolute position of the encoder is arbitrary, so it won't be zero when the system is zero. This means you need a constant representing the offset of the encoder. Besides, the encoder (however unlikely) might jump a little bit and we'd need to change the offset.

Reply:
8cb92cd